feat(key-manager): add documentation for GA (#4169)
scaleway/docs-content: identity-and-access-management/key-manager/api-cli/manage-keys-with-tink.mdx

    ...
    ...
    ...
    ```

## Associated Data

Associated Data (AD) is not encrypted, but it is authenticated. It must be the same when you encrypt and decrypt data, otherwise the decryption fails. This is useful to prevent reading the wrong data in the wrong context. In the table above, the data in both rows 42 and 43 is protected by the same DEK. If we swapped the data, an application would be able to decrypt the data from another row. But, by providing the intended ID as the associated data, the decryption would fail.

### Encrypt data with AD before inserting it

Run the following command to encrypt your data with `Associated Data`. In the example below, associated data like `id42` and `id43` is used to ensure that data from row 42 cannot be decrypted in the context of row 43.

    ```go
    handle, _ := keyset.ReadWithAssociatedData(r, kekAEAD, dekAD)
    primitive, _ := aead.New(handle)  // Same DEK for the two payloads

    secret1, _ := primitive.Encrypt([]byte("This is a secret"), []byte("id42"))
    // Insert secret1 into row 42

    secret2, _ := primitive.Encrypt([]byte("This is another secret"), []byte("id43"))
    // Insert secret2 into row 43
    ```

> Associated Data does not need to be stored, as it can be infered from the context at decryption time. It is also possible to use a unique DEK for each payload. We recommend using Associated Data.

Reviewer: ```suggestion
Associated Data does not need to be stored, as it can be inferred from the context at decryption time. It is also possible to use a unique DEK for each payload. We recommend using Associated Data.
```

## Hierarchy of keys

Unlike KEKs that reside and are managed by Key Manager, DEKs are free: you can generate and have as many as you want.

However, your application still needs to call the Key Manager API:

 - At least once to encrypt a newly generated DEK before storing it, and
 - Each time a DEK needs to be decrypted before use

> Thus, you can use a hierarchy of keys to minimize calls to the Key Manager API (or any remote key mangement service), which can slow down your application and incur charges.

Reviewer: ```suggestion
Thus, you can use a hierarchy of keys to minimize calls to the Key Manager API (or any remote key management service), which can slow down your application and incur charges.
```

masterKeyHandle, _ := keyset.ReadWithAssociatedData(masterKeyFile, kekAEAD, []byte("master-key-001")) // Call the API
masterKeyAEAD, _ := aead.New(masterKeyHandle)

// DEK #1 encrypted and decrypted by the Master Key, rather than by the remote KEK directly. This avoids frequent API calls.
dek1File, _ := os.Open("/path/to/encrypted_dek1.tink")
dekHandle1, _ := keyset.ReadWithAssociatedData(dek1File, masterKeyAEAD, []byte("dek1"))  // No call to the API

// DEK #2 encrypted and decrypted by the Master Key, rather than by the remote KEK directly. This avoids frequent API calls.
dek2File, _ := os.Open("/path/to/encrypted_dek2.tink")
dekHandle2, _ := keyset.ReadWithAssociatedData(dek2File, masterKeyAEAD, []byte("dek2")) // No call to the API

// DEKs are used to encrypt and decrypt the actual data

dek1AEAD, _ := aead.New(dekHandle1)
dek2AEAD, _ := aead.New(dekHandle2)

ct1, _ := dek1AEAD.Encrypt([]byte("this is a secret"), []byte("id42"))
ct2, _ := dek2AEAD.Encrypt([]byte("this is another secret"), []byte("id43"))

/* ... */
/* ... */
```

<Message type="tip">
 Your DEK and KEK do not need to use the same algorithm and cipher.
</Message>

> The example above can work for most use cases. However, there is no "one fits all" approach to create the right key hierarchy. It is up to you to decide on a hierarchy that suits you best,according to your application needs and constraints.

Reviewer: ```suggestion
The example above can work for most use cases. However, there is no "one fits all" approach when creating the right key hierarchy. It is up to you to decide on a hierarchy that suits you best, according to your application needs and constraints.
```
